linux/NvidiaJetson.c

```diff
@@ -0,0 +1,321 @@
+   if (!dir)
+      return;
+
+   while ((entry = readdir(dir)) != NULL) {
+      const char* name = entry->d_name;
+
+      if (name[0] == '.' && (name[1] == '\0' || (name[1] == '.' && name[2] == '\0')))
+         continue;
+
+      if (IsJetsonOrinGPU(name) || IsJetsonXavierGPU(name)) {
+         xSnprintf(GPU_FREQ_SENSOR_PATH, sizeof(GPU_FREQ_SENSOR_PATH), "/sys/class/devfreq/%s/cur_freq", name);
+         xSnprintf(GPU_LOAD_SENSOR_PATH, sizeof(GPU_LOAD_SENSOR_PATH), "/sys/class/devfreq/%s/device/load", name);
+         break;
+      }
+   }
+   closedir(dir);
+}
+
+
+static void NvidiaJetson_GoThroughThermalZones(void) {
+   char path[64], content[4];
+   const struct dirent* entry;
+
+   DIR* dir = opendir("/sys/devices/virtual/thermal");
+   if (!dir)
+      return;
+
+   while ((entry = readdir(dir)) != NULL && (CPU_TEMP_SENSOR_PATH[0] == 0 || GPU_TEMP_SENSOR_PATH[0] == 0)) {
+      const char* name = entry->d_name;
+      ssize_t ret;
+
+      if (String_eq(name, ".") || String_eq(name, ".."))
+         continue;
+
+      if (!String_startsWith(name, "thermal_zone"))
+         continue;
+
+      xSnprintf(path, sizeof(path), "/sys/devices/virtual/thermal/%s/type", name);
+      ret = xReadfile(path, content, sizeof(content));
+      if (ret <= 0)
+         continue;
+
+      content[0] = tolower(content[0]);
+      content[1] = tolower(content[1]);
+      content[2] = tolower(content[2]);
+      content[3] = tolower(content[3]);
+
+      if (CPU_TEMP_SENSOR_PATH[0] == 0 && String_startsWith(content, "cpu")) {
+         xSnprintf(CPU_TEMP_SENSOR_PATH, sizeof(CPU_TEMP_SENSOR_PATH), "/sys/devices/virtual/thermal/%s/temp", name);
+      }
+      if (GPU_TEMP_SENSOR_PATH[0] == 0 && String_startsWith(content, "gpu")) {
+         xSnprintf(GPU_TEMP_SENSOR_PATH, sizeof(GPU_TEMP_SENSOR_PATH), "/sys/devices/virtual/thermal/%s/temp", name);
+      }
+   }
+   closedir(dir);
+}
+
+
+void NvidiaJetson_FindSensors(void) {
+   NvidiaJetson_GoThroughThermalZones();
+   NvidiaJetson_FindGPUDevice();
+}
+
+
+void NvidiaJetson_getCPUTemperatures(CPUData* cpus, unsigned int existingCPUs) {
+   char buffer[22];
+   double temp = xReadNumberFile(CPU_TEMP_SENSOR_PATH, buffer, sizeof(buffer)) / 1000.0;
+   for (unsigned int i=0; i<=existingCPUs; ++i)
+      cpus[i].temperature = temp;
+}
+
+
+enum JetsonValues {
+   JETSON_GPU_LOAD = 0,
+   JETSON_GPU_TEMP = 1,
+   JETSON_GPU_FREQ = 2,
+   JETSON_GPU_TOTAL_COUNT,
+};
+
+
+static void JetsonGPUMeter_updateValues(Meter* this) {
+   char buffer[22];
+   this->values[JETSON_GPU_LOAD] = xReadNumberFile(GPU_LOAD_SENSOR_PATH, buffer, sizeof(buffer));
+   this->curItems = 1; /* only show bar for JETSON_GPU_LOAD */
+
+   this->values[JETSON_GPU_TEMP] = xReadNumberFile(GPU_TEMP_SENSOR_PATH, buffer, sizeof(buffer)) / 1000.0;
+   this->values[JETSON_GPU_FREQ] = xReadNumberFile(GPU_FREQ_SENSOR_PATH, buffer, sizeof(buffer)) / 1000000.0;
+   double percent = this->values[0] / 10.0;
```

Comment:
Inconsistent use of indices into `this->values`; probably meant `JETSON_GPU_LOAD`?